File: leetcode/380.go
type RandomizedSet struct {
	nums    []int
	indices map[int]int
}

func Constructor() RandomizedSet {
	return RandomizedSet{
		indices: make(map[int]int),
	}
}

func (this *RandomizedSet) Insert(val int) bool {
	if _, ok := this.indices[val]; ok {
		return false
	}
	this.nums = append(this.nums, val)
	this.indices[val] = len(this.nums) - 1
	return true
}

func (this *RandomizedSet) Remove(val int) bool {
	idx, ok := this.indices[val]
	if !ok {
		return false
	}
	last := len(this.nums) - 1
	this.nums[idx] = this.nums[last]
	this.indices[this.nums[idx]] = idx
	delete(this.indices, val)
	this.nums = this.nums[:last]
	return true
}

func (this *RandomizedSet) GetRandom() int {
	return this.nums[rand.Intn(len(this.nums))]
}

/**
* Your RandomizedSet object will be instantiated and called as such:
* obj := Constructor();
* param_1 := obj.Insert(val);
* param_2 := obj.Remove(val);
* param_3 := obj.GetRandom();
 */

File: leetcode/380.cpp
class RandomizedSet {
private:
    unordered_map<int,int> indices;
    vector<int> nums;
public:
    RandomizedSet() {

    }
    
    bool insert(int val) {
        if(indices.find(val)!=indices.end()) return false;
        nums.push_back(val);
        indices[val]=nums.size()-1;
        return true;
    }
    
    bool remove(int val) {
        if(indices.find(val)==indices.end()) return false;
        int index = indices[val];
        nums[index]=nums.back();
        indices[nums.back()]=index;
        nums.pop_back();
        indices.erase(val);
        return true;
    }
    
    int getRandom() {
        return nums[rand() % nums.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */